openshift-kube-apiserver/filters/deprecatedapirequest/controller.go

```diff
@@ -0,0 +1,169 @@
+	return &controller{
+		client:       apiclientv1.NewForConfigOrDie(config).DeprecatedAPIRequests(),
+		nodeName:     node,
+		updatePeriod: 10 * time.Second,
+		incoming:     make(chan requestRecord, 100),
+	}
+}
+
+// Controller logs deprecated api requests
+type Controller interface {
+	IsDeprecated(resource, version, group string) bool
+	LogRequest(resource, version, group, verb, user string, timestamp time.Time)
+	Start(stop <-chan struct{})
+}
+
+// controller logs deprecated api requests periodically.
+type controller struct {
+	client       apiclientv1.DeprecatedAPIRequestInterface
+	nodeName     string
+	updatePeriod time.Duration
+	incoming     chan requestRecord
+}
+
+type requestRecord struct {
+	resource  string
+	user      string
+	verb      string
+	timestamp time.Time
+}
+
+// Deprecated return true if the resource is deprecated.
+func (c *controller) IsDeprecated(resource, version, group string) bool {
+	_, ok := deprecatedApiRemovedRelease[schema.GroupVersionResource{
+		Group:    group,
+		Version:  version,
+		Resource: resource,
+	}]
+	return ok
+}
+
+// removedRelease of a specified resource.version.group.
+func (c *controller) removedRelease(r string) string {
+	s := strings.SplitN(r, ".", 3)
+	gvr := schema.GroupVersionResource{Resource: s[0], Version: s[1]}
+	if len(s) == 3 {
+		gvr.Group = s[2]
+	}
+	return deprecatedApiRemovedRelease[gvr]
+}
+
+// LogRequest queues an api request for logging
+func (c *controller) LogRequest(resource, version, group, verb, user string, ts time.Time) {
+	r := resource + "." + version
+	if len(group) > 0 {
+		r += "." + group
+	}
+	select {
+	case c.incoming <- requestRecord{resource: r, user: user, verb: verb, timestamp: ts}:
+	}
+}
+
+// Start the controller
+func (c *controller) Start(stop <-chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stop
+		klog.V(1).Infof("Shutting down deprecatedapirequest controller.")
+		cancel()
+	}()
+
+	//
+	apiRequestCounts := &apiRequestCounts{resources: map[string]*resourceRequestCounts{}}
+
+	// start goroutine that adds incoming log records to the in-memory logs
+	go func() {
+		loaded := sets.NewString()
+		for {
+			select {
+			case rr := <-c.incoming:
+				if !loaded.Has(rr.resource) {
+					apiRequestCounts.IncrementRequestCounts(c.loadOrCreateLogForResource(ctx, rr.resource))
+				}
+				apiRequestCounts.IncrementRequestCount(rr.resource, rr.timestamp, rr.user, rr.verb, 1)
+			case <-stop:
+				return
+			default:
+				// oh, well. best effort (don't want to hang up request chain).
+			}
+		}
+	}()
+
+	// every hour, expire old counts
+	go wait.Until(func() {
+		apiRequestCounts.ExpireOldestCounts(time.Now())
+	}, 1*time.Hour, stop)
+
+	// write out logs every c.updatePeriod
+	go wait.Until(func() {
+		for resource, resources := range apiRequestCounts.resources {
+			_, _, err := v1helpers.UpdateStatus(ctx, c.client, resource, SetRequestCountsForNode(c.nodeName, resources))
+			if err != nil {
```

Comment:
am I just missing where you clear the resourceRequestCounts on successful updates?

Once you update for https://github.com/openshift/kubernetes/pull/592/files#r610869459, the error case will need to call the `.Add` function you've already created to add back the resources we couldn't update for.